File: 0_1.baekjoon/shortest_path/java/src/BOJ11657.java

import java.util.*;
import java.io.*;

public class BOJ11657 {

    private static final long INF = 9999999999L;
    private static int N, M;
    private static Map<Integer, List<int[]>> graph;
    private static long[] dist;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        dist = new long[N + 1];
        graph = new HashMap<>();

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int from = Integer.parseInt(st.nextToken());
            int to = Integer.parseInt(st.nextToken());
            int weight = Integer.parseInt(st.nextToken());

            graph.putIfAbsent(from, new ArrayList<>());
            graph.get(from).add(new int[]{to, weight});
        }

        System.out.println(solve());
    }

    private static String solve() {
        if (hasNegativeCycle()) {
            return "-1";
        }

        StringBuilder sb = new StringBuilder();

        for (int i = 2; i <= N; i++) {
            sb.append(dist[i] == INF ? -1 : dist[i]).append("\n");
        }

        return sb.toString();
    }

    private static boolean hasNegativeCycle() {
        return bellManFord(1);
    }

    private static boolean bellManFord(int start) {
        Arrays.fill(dist, INF);
        dist[start] = 0L;

        for (int i = 0; i < N; i++) {
            for (int node = 1; node <= N; node++) {
                if (dist[node] == INF) {
                    continue;
                }

                for (int[] edge : graph.getOrDefault(node, new ArrayList<>())) {
                    long alt = dist[node] + edge[1];

                    if (dist[edge[0]] > alt) {
                        dist[edge[0]] = alt;

                        if (i == N - 1) {
                            return true;
                        }
                    }
                }
            }
        }

        return false;
    }
}

File: 0_1.baekjoon/shortest_path/kotlin/src/BOJ11657.kt
import java.util.*

private const val INF = 9999999999L
private var N = 0
private var M = 0
private lateinit var graph: MutableList<Edge>
private lateinit var dist: LongArray

data class Edge(val from: Int, val to: Int, val weight: Long)
fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readLine().split(" ").map { it.toInt() }
    N = n
    M = m
    graph = mutableListOf()
    dist = LongArray(N + 1)
    repeat(M) {
        val (from, to, weight) = readLine().split(" ").map { it.toInt() }
        graph.add(Edge(from, to, weight.toLong()))
    }

    print(solve())
}

private fun solve(): String {
    if (hasNegativeCycle(1)) {
        return "-1"
    }
    val sb = StringBuilder()
    for (i in 2..N) {
        if (dist[i] == INF) {
            sb.append(-1)
        } else {
            sb.append(dist[i])
        }
        sb.append("\n")
    }
    return sb.toString()
}

fun hasNegativeCycle(start: Int): Boolean {
    return bellmanFord(start)
}

fun bellmanFord(start: Int): Boolean {
    Arrays.fill(dist, INF)
    dist[start] = 0

    for (i in 0 until N) {
        for (j in 0 until M) {
            val edge = graph[j]
            if (dist[edge.from] == INF) continue
            if (dist[edge.to] > dist[edge.from] + edge.weight) {
                dist[edge.to] = dist[edge.from] + edge.weight

                if (i == N - 1) {
                    return true
                }
            }
        }
    }
    return false
}
